### PR TITLE
PLAT-374: return error result upon outoing call isolation violation

### PR DESCRIPTION
**Pull request by sexyhamster**
**- What I did**
- marshalled error result for outgoing call isolation violation

**- How I did it**

**- How to verify it**

### CODE REVIEW

File: ledger-core/virtual/integration/method_test.go [hunk 1]
```diff
@@ -975,7 +977,177 @@ func TestVirtual_CallMethodFromConstructor_Unordered(t *testing.T) {
...
+
+	Method_PrepareObject(ctx, server, payload.Ready, objectAGlobal)
+
+	outgoingCallRef := gen.UniqueGlobalRef()
+
+	outgoingCall := execution.NewRPCBuilder(outgoingCallRef, objectAGlobal).CallMethod(objectAGlobal, class, "Bar", []byte("123")).SetInterference(contract.CallTolerable)
+	objectAExecutionMock := runnerMock.AddExecutionMock("Foo")
+	objectAExecutionMock.AddStart(
+		func(ctx execution.Context) {
+			require.Equal(t, objectAGlobal, ctx.Request.Callee)
```
Comment (ElSacramento) [inline ledger-core/virtual/integration/method_test.go:1016]: check CallFlags and CTMethod, please

File: ledger-core/virtual/integration/method_test.go [hunk 2]
```diff
@@ -975,7 +977,177 @@ func TestVirtual_CallMethodFromConstructor_Unordered(t *testing.T) {
...
+		&execution.Update{
+			Type:     execution.OutgoingCall,
+			Error:    nil,
+			Outgoing: outgoingCall,
+		},
+	).AddAbort(func() {
+	})
+
+	expectedResult, err := foundation.MarshalMethodErrorResult(throw.E("interference violation: ordered call from unordered call"))
+	if err != nil {
```
Comment (ElSacramento) [inline ledger-core/virtual/integration/method_test.go:1028]: require.Nil

File: ledger-core/virtual/integration/method_test.go [hunk 3]
```diff
@@ -975,7 +977,177 @@ func TestVirtual_CallMethodFromConstructor_Unordered(t *testing.T) {
...
+	}
+
+	msg := utils.NewRequestWrapper(server.GetPulse().PulseNumber, &pl).SetSender(server.JetCoordinatorMock.Me()).Finalize()
+	beforeCount := server.PublisherMock.GetCount()
+	server.SendMessage(ctx, msg)
+	if !server.PublisherMock.WaitCount(beforeCount+1, 10*time.Second) {
+		t.Fatal("failed to wait until all messages returned")
+	}
+
+	server.WaitActiveThenIdleConveyor()
```
Comment (ElSacramento) [inline ledger-core/virtual/integration/method_test.go:1064]:
```
	// wait for all calls and SMs
	{
		testutils.WaitSignalsTimed(t, 20*time.Second, executeDone)
		testutils.WaitSignalsTimed(t, 20*time.Second, server.Journal.WaitAllAsyncCallsDone())
	}
```

File: ledger-core/virtual/integration/method_test.go [hunk 4]
```diff
@@ -975,7 +977,177 @@ func TestVirtual_CallMethodFromConstructor_Unordered(t *testing.T) {
...
+		CallFlags:           payload.BuildCallFlags(contract.CallTolerable, contract.CallDirty),
+		Caller:              server.GlobalCaller(),
+		Callee:              objectAGlobal,
+		CallSiteDeclaration: class,
+		CallSiteMethod:      "Foo",
+		CallOutgoing:        server.RandomLocalWithPulse(),
+		Arguments:           insolar.MustSerialize([]interface{}{}),
+	}
+
+	msg := utils.NewRequestWrapper(server.GetPulse().PulseNumber, &pl).SetSender(server.JetCoordinatorMock.Me()).Finalize()
```
Comment (ElSacramento) [inline ledger-core/virtual/integration/method_test.go:1057]:
```
msg := server.WrapPayload(&pl).Finalize()
```

File: ledger-core/virtual/integration/method_test.go [hunk 5]
```diff
@@ -975,7 +977,177 @@ func TestVirtual_CallMethodFromConstructor_Unordered(t *testing.T) {
...
+	typedChecker := server.PublisherMock.SetTypedChecker(ctx, mc, server)
+	typedChecker.VCallResult.Set(func(res *payload.VCallResult) bool {
+		switch res.Callee {
+		case objectAGlobal:
+			assert.Equal(t, expectedResult, res.ReturnArguments)
+		default:
+			assert.Fail(t, "unexpected VCallResult")
+		}
+		// we should resend that message only if it's CallResult from B to A
+		return false
```
Comment (ElSacramento) [inline ledger-core/virtual/integration/method_test.go:1043]: wrong comment I think

File: ledger-core/virtual/integration/method_test.go (hunk 1 above)
Comment (sexyhamster) [reply ledger-core/virtual/integration/method_test.go:1016]: I'll add check for ctx.Object, we don't have CTMethod inside execution.Context

File: ledger-core/virtual/integration/method_test.go (hunk 2 above)
Comment (sexyhamster) [reply ledger-core/virtual/integration/method_test.go]: fixed with require.NoError

File: ledger-core/virtual/integration/method_test.go (hunk 3 above)
Comment (sexyhamster) [reply ledger-core/virtual/integration/method_test.go]: fixed

File: ledger-core/virtual/integration/method_test.go (hunk 5 above)
Comment (sexyhamster) [reply ledger-core/virtual/integration/method_test.go:1053]: removed :)

File: ledger-core/virtual/integration/method_test.go (hunk 4 above)
Comment (sexyhamster) [reply ledger-core/virtual/integration/method_test.go]: switched to server.SendPayload()

File: ledger-core/virtual/integration/method_test.go (hunk 1 above)
Comment (sexyhamster) [reply ledger-core/virtual/integration/method_test.go]: CallType affects building ctx.Object, so if it is correct, then we have proper payload CallType

File: ledger-core/virtual/integration/method_test.go [hunk 6]
```diff
@@ -975,6 +977,201 @@ func TestVirtual_CallMethodFromConstructor_Unordered(t *testing.T) {
...
+
+	outgoingCall := execution.NewRPCBuilder(outgoingCallRef, objectAGlobal).CallMethod(objectAGlobal, class, "Bar", []byte("123")).SetInterference(contract.CallTolerable)
+	objectAExecutionMock := runnerMock.AddExecutionMock("Foo")
+	objectAExecutionMock.AddStart(
+		func(ctx execution.Context) {
+			assert.Equal(t, objectAGlobal, ctx.Object)
+			assert.Equal(t, contract.MethodIsolation{
+				Interference: contract.CallIntolerable,
+				State:        contract.CallDirty,
+			}, ctx.Isolation)
```
Comment (ElSacramento) [inline ledger-core/virtual/integration/method_test.go:1027]:
```
assert.Equal(t, flags, ctx.Isolation)
```

File: ledger-core/virtual/integration/method_test.go [hunk 7]
```diff
@@ -975,6 +977,201 @@ func TestVirtual_CallMethodFromConstructor_Unordered(t *testing.T) {
...
+	beforeCount := server.PublisherMock.GetCount()
+	server.SendPayload(ctx, &pl)
+	if !server.PublisherMock.WaitCount(beforeCount+1, 10*time.Second) {
+		t.Fatal("failed to wait until all messages returned")
+	}
+
+	{
+		testutils.WaitSignalsTimed(t, 20*time.Second, executeDone)
+		testutils.WaitSignalsTimed(t, 20*time.Second, server.Journal.WaitAllAsyncCallsDone())
+	}
```
Comment (ElSacramento) [inline ledger-core/virtual/integration/method_test.go:1076]:
```suggestion
	{
		testutils.WaitSignalsTimed(t, 20*time.Second, executeDone)
		testutils.WaitSignalsTimed(t, 20*time.Second, server.Journal.WaitAllAsyncCallsDone())
	}
```

Just two WaitSignals

File: ledger-core/virtual/integration/method_test.go [hunk 8]
```diff
@@ -975,6 +977,201 @@ func TestVirtual_CallMethodFromConstructor_Unordered(t *testing.T) {
...
+	require.NoError(t, err)
+
+	expectedResult := []byte("finish A.Foo")
+
+	outgoingCall := execution.NewRPCBuilder(outgoingCallRef, objectAGlobal).CallConstructor(class, "Bar", []byte("123"))
+	objectAExecutionMock := runnerMock.AddExecutionMock("Foo")
+	objectAExecutionMock.
+		AddStart(
+			func(ctx execution.Context) {
+				require.Equal(t, objectAGlobal, ctx.Request.Callee)
```
Comment (ElSacramento) [inline ledger-core/virtual/integration/method_test.go:1123]: isolation flags check

File: ledger-core/virtual/integration/method_test.go [hunk 9]
```diff
@@ -975,6 +977,201 @@ func TestVirtual_CallMethodFromConstructor_Unordered(t *testing.T) {
...
+
+	typedChecker := server.PublisherMock.SetTypedChecker(ctx, mc, server)
+	typedChecker.VCallResult.Set(func(res *payload.VCallResult) bool {
+		switch res.Callee {
+		case objectAGlobal:
+			assert.Equal(t, expectedResult, res.ReturnArguments)
+		default:
+			assert.Fail(t, "unexpected VCallResult")
+		}
+		// we should resend that message only if it's CallResult from B to A
```
Comment (ElSacramento) [inline ledger-core/virtual/integration/method_test.go:1150]: remove

File: ledger-core/virtual/integration/method_test.go (hunk 6 above)
Comment (sexyhamster) [reply ledger-core/virtual/integration/method_test.go]: fixed

File: ledger-core/virtual/integration/method_test.go (hunk 7 above)
Comment (sexyhamster) [reply ledger-core/virtual/integration/method_test.go]: fixed

File: ledger-core/virtual/integration/method_test.go (hunk 9 above)
Comment (sexyhamster) [reply ledger-core/virtual/integration/method_test.go]: fixed

File: ledger-core/virtual/integration/method_test.go (hunk 8 above)
Comment (sexyhamster) [reply ledger-core/virtual/integration/method_test.go]: fixed

File: ledger-core/virtual/integration/method_test.go [hunk 10]
```diff
@@ -975,6 +977,188 @@ func TestVirtual_CallMethodFromConstructor_Unordered(t *testing.T) {
...
+			assert.Equal(t, expectedResult, res.ReturnArguments)
+		default:
+			assert.Fail(t, "unexpected VCallResult")
+		}
+		return false
+	}).ExpectedCount(1)
+
+	pl := payload.VCallRequest{
+		CallType:            payload.CTMethod,
+		CallFlags:           payload.BuildCallFlags(contract.CallTolerable, contract.CallDirty),
```
Comment (ElSacramento) [inline ledger-core/virtual/integration/method_test.go:1148]: Dou you mean to test isolation negotiation for A.Foo?  
```
flags = contract.MethodIsolation{Interference: contract.CallIntolerable, ...
```

File: ledger-core/virtual/integration/method_test.go [hunk 11]
```diff
@@ -975,6 +977,188 @@ func TestVirtual_CallMethodFromConstructor_Unordered(t *testing.T) {
...
+	typedChecker := server.PublisherMock.SetTypedChecker(ctx, mc, server)
+	typedChecker.VCallResult.Set(func(res *payload.VCallResult) bool {
+		switch res.Callee {
+		case objectAGlobal:
+			assert.Equal(t, expectedResult, res.ReturnArguments)
+		default:
+			assert.Fail(t, "unexpected VCallResult")
+		}
+		return false
+	}).ExpectedCount(1)
```
[the remark this answers is not in the document]
Comment (sexyhamster) [reply ledger-core/virtual/integration/method_test.go]: fixed

File: ledger-core/virtual/integration/method_test.go (hunk 10 above)
Comment (sexyhamster) [reply ledger-core/virtual/integration/method_test.go]: my bad, fixed

**Pull request merged**